tests/e2e/restart_test.go

```diff
@@ -137,23 +157,183 @@ func (s *RestartTestData) TestPluginKill(c *C) {
+func (s *RestartTestData) TestRecoverMountsAfterRestart(c *C) {
+	misc.LogTestStart(c.TestName())
+
+	c.Skip("Skipping verification of attached status after docker restart")
+	// 1. Run containers with restart-always flag
+	out, err := dockercli.AttachVolumeWithRestart(s.config.DockerHosts[1], s.volumeName, s.containerName[0])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// Run second container 
+	out, err = dockercli.AttachVolumeWithRestart(s.config.DockerHosts[1], s.volumeName, s.containerName[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 2. Restart docker
+	out, err = dockercli.RestartDocker(s.config.DockerHosts[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	misc.SleepForSec(60)
+
+	// 3. Verify the volume is attached to the VM
+	status, err := dockercli.GetVolumeStatus(s.config.DockerHosts[1], s.volumeName)
+	c.Assert(err, IsNil, Commentf("Failed to fetch status for volume %s", s.volumeName))
+	c.Assert(s.config.DockerHostNames[0], Equals, status["attached_vm"], Commentf("Volume is not attached to VM %s, attached to %s",
+		s.config.DockerHostNames[0], status["attached_vm"]))
+
+	// 4. Run third container
+	out, err = dockercli.AttachVolumeWithRestart(s.config.DockerHosts[1], s.volumeName, s.containerName[2])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 5. Stop the three containers
+	out, err = dockercli.RemoveContainer(s.config.DockerHosts[1], s.containerName[0])
+	c.Assert(err, IsNil, Commentf(out))
+
+	out, err = dockercli.RemoveContainer(s.config.DockerHosts[1], s.containerName[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	out, err = dockercli.RemoveContainer(s.config.DockerHosts[1], s.containerName[2])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 6. Run container on other host to verify the volume is detached after stopping the containers
+	out, err = dockercli.ExecContainer(s.config.DockerHosts[0], s.volumeName, s.containerName[0])
+	c.Assert(err, IsNil, Commentf(out))
+
+	misc.LogTestEnd(c.TestName())
+}
+
+// TestLongVolumeName - Verifies volumes can be created on the default
+// datastore by long and short names.
+// 1. Start container with restart flag and use short name for volume (vol1)
+// 2. Restart docker
+// 3. Start another instance of container with short name for volume (vol1)
+// 4. Get the volume properties and figure the datastore of the volume
+// 4. Start another instance of container with long name for volume (vol1@<datastore-name>)
+// 5. Stop all 3 containers and confirm disk is detached (verify by exec'ing a container on another host).
+func (s *RestartTestData) TestLongVolumeName(c *C) {
+	misc.LogTestStart(c.TestName())
+
+	// 1. Run container with restart-always flag
+	out, err := dockercli.AttachVolumeWithRestart(s.config.DockerHosts[1], s.volumeName, s.containerName[0])
+	c.Assert(err, IsNil, Commentf(out))
+
+	// 2. Restart docker
+	out, err = dockercli.RestartDocker(s.config.DockerHosts[1])
+	c.Assert(err, IsNil, Commentf(out))
+
+	misc.SleepForSec(60)
```

Comment:
Without the sleep (this may need documentation or a rethink on refcount init), the remove container calls translate to unmount requests from docker to the plugin. The plugin is not ready yet so we mark dirty on the refcount and the unmount returns immediately.

The refcounter in parallel may or may not have found the running containers and can keep a reference count on the volume causing it to remain attached to the VM. Which is why the earlier CI failed (without the sleep).

At least unmount should fail the call to docker if the refcounter is still not initialized perhaps.